flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -191,7 +191,7 @@
 
 	private final ResultPartition[] producedPartitions;
 
-	private final SingleInputGate[] inputGates;
+	private final InputGate[] inputGates;
```

Comment:
I would keep it as `SingleInputGate[]` to avoid that copying loop `SingleInputGate[]` -> `InputGate[]` in Task constructor at the moment.

Later, we could move `inputGatesById` completely to `NetworkEnviroment`, when we refactor:
- `TaskExecutor.updatePartitions` to use `NetworkEnviroment.update(PartitionInfo)`
- try to move `triggerPartitionProducerStateCheck` to `SingleInputGate` and pass `partitionProducerStateChecker` to `SingleInputGate` instead of having it in `TaskActions`

`inputGatesById` looks like the last blocker to change `SingleInputGate[]` -> `InputGate[]` in Task.

Reply:
I agree with the following refactors for `updatePartitions` and `triggerPartitionProducerStateCheck` which could help remove `inputGatesById` in task.

But the change of `SingleInputGate[]` to `InputGate[]` is necessary here because we have to loop the `SingleInputGate[]` from `NetworkEnvironment#createInputGates` to generate `InputGateWithMetrics` which implements `InputGate`. The task would finally maintain `InputGate[]` instead of `SingleInputGate`, so this change would be consistent with future plans.

Did I misunderstand your concern?

Reply:
Oh, sorry, you are right, too fast review :)